preview/controllers.py

```diff
@@ -1,18 +1,159 @@
+    except store.DoesNotExist as e:
+        raise NotFound('No preview available') from e
+
+    data = {'added': metadata.added,
+            'checksum': metadata.checksum,
+            'size_bytes': metadata.size_bytes}
+    headers = {'ETag': metadata.checksum}
+    return data, HTTPStatus.OK, headers
+
+
+def get_preview_content(source_id: str, checksum: str,
+                        none_match: Optional[str] = None) -> Response:
+    """
+    Handle request for preview content.
+
+    Parameters
+    ----------
+    source_id : str
+        Unique identifier for the source package.
+    checksum : str
+        State of the source package to which this preview corresponds.
+    none_match : str or None
+        If not None, will return 304 Not Modified if the current preview
+        checksum matches.
+
+    Returns
+    -------
+    io.BytesIO
+        Stream containing the preview content.
+    int
+        HTTP status code.
+    dict
+        Headers to add to the response.
+
+    """
+    st = store.PreviewStore.current_session()
+    try:
+        if none_match is not None:
+            preview_checksum = st.get_preview_checksum(source_id, checksum)
+            if none_match == preview_checksum:
+                headers = {'ETag': preview_checksum}
+                return None, HTTPStatus.NOT_MODIFIED, headers
+
+        preview = st.get_preview(source_id, checksum)
+    except store.DoesNotExist as e:
+        raise NotFound('No preview available') from e
+
+    if preview.metadata is None or preview.content is None:
+        raise InternalServerError('Unexpected error loading content')
+
+    headers = {'ETag': preview.metadata.checksum,
+               'Content-type': 'application/pdf',}
+    return preview.content.stream, HTTPStatus.OK, headers
+
+
+def deposit_preview(source_id: str, checksum: str, stream: IO[bytes],
+                    content_type: Optional[str],
+                    overwrite: bool = False) -> Response:
+    """
+    Handle a request to deposit the content of a preview.
+
+    Parameters
+    ----------
+    source_id : str
+        Unique identifier for the source package.
+    checksum : str
+        State of the source package to which this preview corresponds.
+    stream : io.BytesIO
+        Byte-stream from the request body.
+    content_type : str
+        Value of the ``Content-type`` request header.
+    overwrite : bool
+        Value of the ``Overwrite`` request header.
+
+    Returns
+    -------
+    dict
+        Metadata about the deposit.
+    int
+        HTTP status code.
+    dict
+        Headers to add to the response.
+
+    """
+    if content_type is None or content_type != 'application/pdf':
+        logger.info('Request is missing content-type header')
+        raise BadRequest(f'Invalid content type: {content_type}')
+    st = store.PreviewStore.current_session()
+    preview = Preview(source_id, checksum, content=Content(stream=stream))
```

Comment:
Kind of confusing to have two different checksums running around, sorry about that. Currently the requestor is only passing the checksum of the source package. The reason I didn't also have them pass the checksum of the preview is that it would require the requestor to fully consume the preview before passing it along. That would be kind of a bummer, because the requestor would either have to hold the whole preview in memory or retrieve the preview twice. This way, the requestor can request a PDF from (say) the compiler service, and stream the response data directly into the PUT request to the preview service.

Reply:
Ah, OK, thanks for the clarification.

Reply:
That being said, there is no reason why the source of the PDF (compiler service, file manager service) couldn't provide the checksum for the PDF in a header, which would be available to the requestor at the beginning. So actually we could do this.

Reply:
In fact I believe that they are already providing this via ``ETag``. Will revise.

Reply:
Added in c4c9f75